### ISSUE TITLE
Write a guide in the wiki page

### ISSUE THREAD

**Issue by fredtcaroli** [NNBlocks/NNBlocks#42]

**Comment by fredtcaroli** [comment]
The page is created at https://github.com/NNBlocks/NNBlocks/wiki/User-Guide

**Comment by pasky** [comment]
Thanks for that guide - overally, it's been helpful!

However, as described now in the User Guide, it's not clear at all that running trainer.train() will run just a _single iteration_ of the training algorithm - I got stuck on this for long, wondering why my models weren't learning!  It would be also good to briefly mention what the preferred way to actually train models is right now, given issue #48.

Overally, it would be great to start the guide or even the README with a canonical example that not only builds a model, but also trains it and applies it on some data. Right now, from the README it's not really clear what to do once you build a model.

**Comment by pasky** [comment]
(I did notice now that the user guide mentions "adjust once", but it's a pretty easy single word to overlook. Overally, I think best would be to have a canonical end-to-end example that does everything, features prominently.)

**Comment by pasky** [comment]
A concrete example of an end-to-end code (if I'm doing it right!) that learns a linear separation of two samples:

``` python
I = nnb.InputLayer(ndim=1)
P = nnb.PerceptronLayer(insize=2, outsize=1)
Y = nnb.InputLayer(ndim=1, dtype='int32')
N = I | P
E = (N & Y) | nnb.cost.MeanSquareError()
t = nnb.train.SGDTrainer(model=E, learning_rate=0.1)
tdata = [[np.array([0, 1]), np.array([0], dtype='int32')],
         [np.array([1, 0]), np.array([1], dtype='int32')]]
ts = nnb.train.TrainSupervisor(trainer=t, dataset=tdata, eval_dataset=tdata, epochs_num=200)
ts.train()
ff = N.compile()
for i in range(2):
    print(tdata[i][0], ff(tdata[i][0]), tdata[i][1])
print('W', P.params[0].get_value().T, 'b', P.params[1].get_value())
```

**Comment by fredtcaroli** [comment]
This example looks good!
Right now I'm not developing the guide or making any effort to document the training further because the whole framework will change its ways in the near future. Because of #46 and #48, the new way to train a model will be something along these lines:

``` python
M = make_model()
X1 = nnb.data.Input(get_x1_data(), name='x1')
X2 = nnb.data.Input(get_x2_data(), name='x2')
Y = nnb.data.Input(get_y_data(), name='y')

dataset = nnb.data.Dataset([X1, X2, Y])

trainer = nnb.train.SGDTrainer(model=M, dataset=dataset, batch_size=20)
while not trainer.done_training(): #Not sure yet if this method should exist or how to configure it
    trainer.train()
    do_some_evaluation(M)
```

I'm having some trouble developing the training, because it's hard to take into account the padding in the data while training. I want to make the data issues completely transparent to the user, and this is proving to be hard right now. When I'm done developing this, the training should look like the code above.

**Comment by pasky** [comment]
Thank you, that's been quite helpful. I'm curious about how that'll work, but so far I'm content with using the current training infrastructure (I don't even use GPU for theano so far, though.)  Maybe the most annoying part (after finally getting how it's supposed to work) was that I had to implement my own class-balancing logic (got spoiled by sklearn!).
